plutus-core/plutus-core/src/PlutusCore/Normalize/Internal.hs

```diff
@@ -118,12 +121,44 @@ was previously added to an environment (while handling the function application
 this value and rename all bound variables in it to preserve the global uniqueness condition. It is
 safe to do so, because picked values cannot contain uninstantiated variables as only normalized types
 are added to environments and normalization instantiates all variables presented in an environment.
+
+See also Note [Normalization of built-in types].
 -}
 
+{- Note [Normalization of built-in types]
+Instantiating a polymorphic built-in type amounts to applying it to some arguments. However,
+the notion of "applying" is ambiguous, it can mean one of these two things:
+
+1. lifting the built-in type to 'Type' and applying that via 'TyApp'
+2. applying the built-in type right inside the universe to get a monomorphized type tag
+   (e.g. the default universe has 'DefaultUniApply' for that purpose)
+
+We need both of these things. The former allows us to assign types to polymorphic built-in functions
+(otherwise applying a built-in type to a type variable would be unrepresentable), the latter is
+used at runtime to juggle type tags so that we can avoid @unsafeCoerce@-ing, bring instances in
+scope via 'bring' etc -- for all of that we have to have fully monomorphized type tags at runtime.
+
+So in order for type checking to work we need to normalize polymorphic built-in types. For that
+we simply turn intra-universe applications into regular type applications during type normalization.
+-}
+
+-- See Note [Normalization of built-in types].
+-- | Normalize a built-in type by replacing each application inside the universe with regular
+-- type application.
+normalizeUni :: forall k (a :: k) uni tyname. HasUniApply uni => uni (T a) -> Type tyname uni ()
```

Comment:
> This feels a little odd to me, since in some ways this seems like the opposite of normalization, if normalization is reduction-like.

Normalization doesn't have to be reduction-like, it's literally a process of reducing something to a normal form, whatever that form is.

> Shouldn't regular type application of builtins "reduce" to application in the built-in universe, if anything?

That would be harder to implement, because collapsing a general `Type` into a `SomeTypeIn` is harder than expanding a `SomeTypeIn` into a `Type`. And it would also be impossible to do in the general case, 'cause you can't collapse an application of a built-in type to, say, a type variable into an intra-universe application as there are no type variables there. So I guess we could "reduce" application of built-in types in some cases and not reduce them in others and still make the whole thing work, but that requires substantially more logic and is also a lot harder to get right.

I'll expand the docs.